sdk/eventhub/event-hubs/src/eventDataBatch.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { EventData, toAmqpMessage } from "./eventData";
+import { ConnectionContext } from "./connectionContext";
+import { AmqpMessage } from "@azure/core-amqp";
+import { message } from "rhea-promise";
+
+/**
+ * A class representing a batch of events which can be passed to the `send` method of a `EventConsumer` instance.
+ * This batch is ensured to be under the maximum size limited supported by Azure Event Hubs service.
+ *
+ * Use the `tryAdd` function on the EventDataBatch to add events in a batch.
+ * @class
+ */
+export class EventDataBatch {
+  /**
+   * @property Describes the amqp connection context for the Client.
+   */
+  private _context: ConnectionContext;
+  /**
+   * @property Total length of events.
+   */
+  private _eventsLength: number;
+  /**
+   * @property A value that is hashed to produce a partition assignment.
+   * It guarantees that messages with the same partitionKey end up in the same partition.
+   * Specifying this will throw an error if the producer was created using a `paritionId`.
+   */
+  private _partitionKey?: string;
+  /**
+   * @property The maximum size allowed for the batch.
+   */
+  private readonly _maxSizeInBytes: number;
+  /**
+   * @property Current size of the batch in bytes.
+   */
+  private _size: number;
+  /**
+   * @property Encoded amqp messages.
+   */
+  private _encodedMessages: Buffer[] = [];
+  /**
+   * @property Encoded batch message.
+   */
+  private _batchMessage: Buffer | undefined;
+
+  /**
+   * @constructor
+   * @internal
+   * @ignore
+   */
+  constructor(context: ConnectionContext, maxSizeInBytes: number, partitionKey?: string) {
+    this._context = context;
+    this._maxSizeInBytes = maxSizeInBytes;
+    this._partitionKey = partitionKey;
+    this._size = 0;
+    this._eventsLength = 0;
+  }
+
+  /**
+   * @property A value that is hashed to produce a partition assignment.
+   * @readonly
+   */
+  get partitionKey(): string | undefined {
+    return this._partitionKey;
+  }
+
+  /**
+   * @property Size of a batch of events.
+   * @readonly
+   */
+  get size(): number | undefined {
+    return this._size;
+  }
+
+   /**
+   * @property Encoded batch message.
+   * @readonly
+   */
+  get batchMessage(): Buffer | undefined {
+    return this._batchMessage;
+  }
+
+  /**
+   * Tries to add an event data to the batch if permitted by the batch's size limit.
+   * @param eventData  An individual event data object.
+   * @returns A boolean value indicating if the event data has been added to the batch or not.
+   */
+  public tryAdd(eventData: EventData): boolean {
+    const currentSize = this.getEventBatchSize(eventData);
+    if (currentSize > this._maxSizeInBytes) {
```

Comment:
The check on whether the current size is under max size or not is being done in 2 places. If size is under limit then certain properties are updated inside `getEventBatchSize` (`this._batchMessage` and `this._encodedMessages`) and others are updated inside `tryAdd` (`this._eventsLength` and `this._size` )

It would be better if all the property updates are done in a single place. I would suggest moving the update of `this._eventsLength` and `this._size` inside `getEventBatchSize`

Reply:
On second thoughts, after the above changes, there will not be much left inside `tryAdd()`, so we might as well consider moving the code from inside `getEventBatchSize` to `tryAdd`